ts/private/ts_project_worker.js

```diff
@@ -363,45 +439,48 @@ function createEmitAndLibCacheAndDiagnosticsProgram(
     return builder;
 }
 
-/** @type {ts.FormatDiagnosticsHost} */
-const formatDiagnosticHost = {
-    getCanonicalFileName: (path) => path,
-    getCurrentDirectory: ts.sys.getCurrentDirectory,
-    getNewLine: () => ts.sys.newLine,
-};
-
-function printDiagnostics(diagnostics) {
-    worker.log('');
-    worker.log(ts.formatDiagnostics(diagnostics, formatDiagnosticHost));
-}
-
-function createProgram(args, initialInputs) {
+function createProgram(args, inputs, output) {
     const {options} = ts.parseCommandLine(args);
     const compilerOptions = {...options}
 
+    compilerOptions.outDir = path.join("__synthetic__outdir__", compilerOptions.outDir);
```

Comment:
this is the biggest culprit there is. under bazel outDir and and rootDir is the same and this creates all sorts of state problems. like refusing to discard cache entries based on simple outDir and rootDir checks.

https://github.com/microsoft/TypeScript/blob/89e928e8b41368bb0888c6d399854c53a2dbf3d4/src/compiler/program.ts#L4601-L4635

Reply:
TODO: same needs to be done for declarationDir as well.